esmvalcore/_recipe.py

```diff
@@ -232,20 +232,26 @@ def _augment(base, update):
 
 def _dataset_to_file(variable, config_user):
     """Find the first file belonging to dataset from variable info."""
-    files = get_input_filelist(
-        variable=variable,
-        rootpath=config_user['rootpath'],
-        drs=config_user['drs'],
-    )
+    files = _get_input_files(variable, config_user)
     if not files and variable.get('derive'):
-        first_required = get_required(variable['short_name'])[0]
-        _augment(first_required, variable)
-        files = get_input_filelist(
-            variable=first_required,
-            rootpath=config_user['rootpath'],
-            drs=config_user['drs'],
-        )
-    check.data_availability(files, variable)
+        required_vars = get_required(variable['short_name'],
+                                     variable['project'])
+        for required_var in required_vars:
+            if not required_var.get('optional'):
+                selected_var = required_var
+                break
+        else:
+            logger.warning(
+                "All required variables for the derivation of '%s' are marked "
+                "as 'optional', this might cause errors due to unavailable "
+                "data", variable['short_name'])
+            selected_var = required_vars[0]
+        _augment(selected_var, variable)
+        _add_cmor_info(selected_var, override=True)
+        files = _get_input_files(selected_var, config_user)
+        check.data_availability(files, selected_var)
+    else:
+        check.data_availability(files, variable)
```

Comment:
I would suggest to write this function as follows, do you think that would work? Because even if all variables are marked as optional, I do not think you can have a successful run if there is no data at all, because then the derivation would result in an empty cube.
```python
def _dataset_to_file(variable, config_user):
    """Find the first file belonging to dataset from variable info."""
    files = _get_input_files(variable, config_user)
    if not files and variable.get('derive'):
        required_vars = get_required(variable['short_name'],
                                     variable['project'])
        for required_var in required_vars:
            _augment(required_var, variable)
            _add_cmor_info(required_var, override=True)
            files = _get_input_files(required_var, config_user)
            if files:
                variable = required_var
                break
    check.data_availability(files, variable)
    return files[0]
```

Reply:
Makes totally sense, yes :+1: